modules/ROOT/pages/dataweave-type-selector.adoc

```diff
@@ -0,0 +1,209 @@
+
+Example: DataWeave script
+[source,dataweave,linenums]
+----
+%dw 2.0
+---
+type User = {name: String}
+type NameType = User.name
+var name: NameType = "Seba"
+---
+typeOf(name)
+----
+
+Output 
+----
+"String"
+----
+
+== Examples on How to Use Type Selector
+
+The following example describes how type selection is useful for declaring complex types from other types already declared:
+
+Example: Composition
+
+DataWeave script
+[source,dataweave,linenums]
+----
+%dw 2.0
+output application/json
+---
+type User = {
+ address : {country: String, city: String, street: String, number: Number},
+ userName : String
+}
+type Address = User.address
+
+var userAddress: Address = {country: "Argentina", city: "Rosario", street: "Calle Falsa", number: 123}
+---
+userAddress
+----
+
+Output
+[source,json,linenums]
+----
+{
+ "country": "Argentina",
+ "city": "Rosario",
+ "street": "Calle Falsa",
+ "number": 123
+}
+----
+
+The next example shows how type selection can also be used with Union Types involving object types:
+
+Example: Union Types
+
+DataWeave script
+[source,dataweave,linenums]
+----
+%dw 2.0
+output application/json
+---
+type ID = {name: String} | {id: Number}
+var name: ID = {name: "Seba"}
+var id: ID = {id: 38123}
+var nameString: ID.name = "Seba"
+var idNumber: ID.id = 38123
+---
+[name, id, nameString ++ " is of type " ++ typeOf(nameString), idNumber ++ " is of type " ++ typeOf(idNumber)]
+----
+
+Output
+[source,json,linenums]
+----
+[
+  {
+    "name": "Seba"
+  },
+  {
+    "id": 38123
+  },
+  "Seba is of type String",
+  "38123 is of type Number"
+]
+----
+
+However, the following script fails because the second type is not an object type and there is no way to reference it with the selector:
+
+DataWeave script
+[source,dataweave,linenums]
+----
+%dw 2.0
+output application/json
+---
+type UnionType = {name: String} | Number
+type Fail = UnionType.name
+---
+{}
+----
+
+Output
+[source,json,linenums]
+----
+Cannot do selection on Type: Number
+
+4| type Fail = UnionType.name
+----
+
+In the following example, Type selection also supports namespaces:
+
+Example: Namespace
+
+DataWeave script
+[source,dataweave,linenums]
+----
+%dw 2.0
+output application/json
+---
+ns ns1 http://acme.com
+type User = {
+ ns1#name : String,
+ name : Number,
+}
+var nameString: User.ns1#name = "Seba"
+var nameNumber: User.name = 123
+---
+{
+   "NameString" : typeOf(nameString),
+   "NameNumber" : typeOf(nameNumber)
+}
+----
+
+Output
+[source,json,linenums]
+----
+{
+ "NameString": "String",
+ "NameNumber": "Number"
+}
+----
+
+This example outlines how Type selection preserves metadata associated with the type:
+
+Example: Metadata
+
+DataWeave script
+[source,dataweave,linenums]
+----
+%dw 2.0
+output application/json
+---
```

Comment:
```suggestion

```
Extra `---` removed. That differentiates the scripts header content (above the dashes) from the body (below). The type definition goes in the header.

Reply:
You can test the script in the playground to see when there's an issue: https://dataweave.mulesoft.com/learn/playground